src/commands/app/pack.js

```diff
@@ -0,0 +1,266 @@
+    await this.config.runHook('pre-pack', { appConfig, artifactsFolder: DEFAULTS.ARTIFACTS_FOLDER })
+
+    // 2. copy files to package phase
+    this.log('Copying files...')
+    const fileList = await this.filesToPack()
+    await this.copyPackageFiles(DEFAULTS.ARTIFACTS_FOLDER, fileList)
+
+    // 3. add/modify artifacts phase
+    this.log('Creating configuration files...')
+    await this.createDeployYamlFile(appConfig)
+    await this.addCodeDownloadAnnotation(appConfig)
+    // doing this before zip so other things can be added to the zip
+    await this.config.runHook('post-pack', { appConfig, artifactsFolder: DEFAULTS.ARTIFACTS_FOLDER })
+
+    // 4. zip package phase
+    this.log(`Zipping package artifacts folder '${DEFAULTS.ARTIFACTS_FOLDER}' to '${flags.output}'...`)
+    await fs.remove(flags.output)
+    await this.zipHelper(DEFAULTS.ARTIFACTS_FOLDER, flags.output)
+
+    this.log('Packaging done.')
+  }
+
+  /**
+   * Creates the deploy.yaml file
+   *
+   * @param {object} appConfig the app's configuration file
+   */
+  async createDeployYamlFile (appConfig) {
+    // get extensions
+    let extensions
+    if (appConfig.implements?.filter(item => item !== 'application').length > 0) {
+      extensions = appConfig.implements.map(ext => ({ extensionPointId: ext }))
+    }
+
+    // get workspaces
+    let workspaces
+    if (appConfig.aio?.project?.workspace?.name) {
+      workspaces = []
+      workspaces.push(appConfig.aio?.project?.workspace?.name)
+    }
+
+    // get apis
+    let apis
+    if (appConfig.aio?.project?.workspace?.details?.services?.length > 0) {
+      apis = appConfig.aio.project.workspace.details.services.map(service => ({ code: service.code }))
+    }
+
+    // read name and version from package.json
+    const application = {
+      id: appConfig.packagejson.name,
+      version: appConfig.packagejson.version
+    }
+
+    let meshConfig
+    // ACNA-2041
+    // get the mesh config by running the `aio api-mesh:get` command (if available)
+    // in the interim, we need to process the output to get the proper json config
+    // TODO: send a PR to their plugin to have a `--json` flag
+    const command = await this.config.findCommand('api-mesh:get')
+    if (command) {
+      const { stdout } = await execa('aio', ['api-mesh', 'get'], { cwd: process.cwd() })
+      // until we get the --json flag, we parse the output
+      const idx = stdout.indexOf('{')
+      meshConfig = JSON.parse(stdout.substring(idx))
+    } else {
+      aioLogger.debug('api-mesh:get command was not found, meshConfig is not available for app:pack')
+    }
+
+    const deployJson = {
+      $schema: 'http://json-schema.org/draft-07/schema',
+      $id: 'https://adobe.io/schemas/app-builder-templates/1',
+      application,
+      extensions,
+      workspaces,
+      apis,
+      meshConfig,
+      runtime: true // always true for App Builder apps
+    }
+
+    await writeFile(
```

Comment:
is it yaml or json? I know yaml includes json, but this looks a little confusing.

Reply:
we write out yaml, as that is what `yaml.dump` outputs, and takes json as input.